registry/crypto/functions.go

```diff
@@ -0,0 +1,449 @@
+		return crt, fmt.Errorf(
+			"error parsing private key: %s",
+			err,
+		)
+	}
+
+	crt.Cert = string(cert)
+	crt.Key = string(key)
+
+	return crt, nil
+}
+
+// GenerateCertificateAuthority generates a certificate authority using the provided common name and validity period.
+//
+// Parameters:
+//   - cn: the common name for the certificate authority
+//   - daysValid: the number of days the certificate authority is valid for
+//
+// Returns:
+//   - Certificate: the generated certificate authority
+//   - error: an error if any occurred during the generation process
+//
+// Example:
+//
+//	{{ generateCertificateAuthority "example.com" 365 }} // Output: {"Cert":"b64cert","Key":"b64key"}
+func (ch *CryptoRegistry) GenerateCertificateAuthority(
+	cn string,
+	daysValid int,
+) (Certificate, error) {
+	priv, err := rsa.GenerateKey(cryptorand.Reader, 2048)
+	if err != nil {
+		return Certificate{}, fmt.Errorf("error generating rsa key: %s", err)
+	}
+
+	return ch.generateCertificateAuthorityWithKeyInternal(cn, daysValid, priv)
+}
+
+// GenerateCertificateAuthorityWithPEMKey generates a certificate authority using the provided common name, validity period, and private key in PEM format.
+//
+// Parameters:
+//   - cn: the common name for the certificate authority
+//   - daysValid: the number of days the certificate authority is valid for
+//   - privPEM: the private key in PEM format
+//
+// Returns:
+//   - Certificate: the generated certificate authority
+//   - error: an error if any occurred during the generation process
+//
+// Example:
+//
+//	{{ generateCertificateAuthorityWithPEMKey "example.com" 365 "privPEM" }} // Output: {"Cert":"b64cert","Key":"b64key"}
+func (ch *CryptoRegistry) GenerateCertificateAuthorityWithPEMKey(
+	cn string,
+	daysValid int,
+	privPEM string,
+) (Certificate, error) {
+	priv, err := ch.parsePrivateKeyPEM(privPEM)
+	if err != nil {
+		return Certificate{}, fmt.Errorf("parsing private key: %s", err)
+	}
+	return ch.generateCertificateAuthorityWithKeyInternal(cn, daysValid, priv)
+}
+
+// GenerateSelfSignedCertificate generates a new, self-signed x509 certificate using a 2048-bit RSA private key.
+//
+// Parameters:
+//   - cn: the common name for the certificate
+//   - ips: a list of IP addresses
+//   - alternateDNS: a list of alternate DNS names
+//   - daysValid: the number of days the certificate is valid for
+//
+// Returns:
+//   - Certificate: the generated certificate
+//   - error: an error if any occurred during the generation process
+//
+// Example:
+//
+//	{{ generateSelfSignedCertificate "example.com" ["127.0.0.1"] ["localhost"] 365 }} // Output: {"Cert":"b64cert","Key":"b64key"}
+func (ch *CryptoRegistry) GenerateSelfSignedCertificate(
+	cn string,
+	ips []interface{},
+	alternateDNS []interface{},
+	daysValid int,
+) (Certificate, error) {
+	priv, err := rsa.GenerateKey(cryptorand.Reader, 2048)
+	if err != nil {
+		return Certificate{}, fmt.Errorf("error generating rsa key: %s", err)
```

Comment:
```suggestion
		return Certificate{}, fmt.Errorf("error generating RSA key: %s", err)
```

Reply:
Due to backward compatibility with Sprig and the lack of an accessible error variable to use with errors.As or errors.Is, most Sprig users who rely on the crypto package end up comparing strings. So, unfortunately, I cannot change it 😢

Also, please consider this discussion: https://github.com/go-sprout/sprout/pull/46#issuecomment-2282901894